seed/seed20.go

```diff
@@ -356,53 +356,189 @@ func (s *seed20) lookupSnap(snapRef naming.SnapRef, optSnap *internal.Snap20, ch
+			// skipped optional snap is ok
+			return nil, errSkipped
+		}
 		return nil, err
 	}
+	seedSnap.Essential = essential
+	seedSnap.Required = required
+	if essential {
+		if sntoc.modelSnap.SnapType == "gadget" {
+			// validity
+			info, err := readInfo(seedSnap.Path, seedSnap.SideInfo)
+			if err != nil {
+				return nil, err
+			}
+			if info.Base != s.model.Base() {
+				return nil, fmt.Errorf("cannot use gadget snap because its base %q is different from model base %q", info.Base, s.model.Base())
+			}
+			// TODO: when we allow extend models for classic
+			// we need to add the gadget base here
+		}
 
-	s.snaps = append(s.snaps, seedSnap)
-	s.modes = append(s.modes, modes)
-
+		seedSnap.EssentialType = snapTypeFromModel(sntoc.modelSnap)
+	}
 	return seedSnap, nil
 }
 
-var errFiltered = errors.New("filtered out")
+func (s *seed20) doLoadMeta(tm timings.Measurer) error {
+	// setup essential snaps cache
+	if s.essCache == nil {
+		// 4 = snapd+base+kernel+gadget
+		s.essCache = make(map[string]*Snap, 4)
+	}
+	cacheEssential := func(snType string, essSnap *Snap) {
+		s.essCacheMu.Lock()
+		defer s.essCacheMu.Unlock()
+		s.essCache[snType] = essSnap
+	}
+	cachedEssential := func(snType string) *Snap {
+		s.essCacheMu.Lock()
+		defer s.essCacheMu.Unlock()
+		return s.essCache[snType]
+	}
+	runMode := []string{"run"}
+
+	n := len(s.snapsToConsiderCh)
+	close(s.snapsToConsiderCh)
+	if n > 0 {
+		s.snaps = make([]*Snap, n)
+		s.modes = make([][]string, n)
+	}
+
+	njobs := s.nLoadMetaJobs
+	if njobs < 1 {
+		njobs = 1
+	}
+	stopCh := make(chan struct{})
+	outcomesCh := make(chan error, njobs)
+	for j := 1; j <= njobs; j++ {
+		jtm := tm.StartSpan(fmt.Sprintf("do-load-meta[%d]", j), fmt.Sprintf("snap metadata loading job #%d", j))
+		go func() {
+		Consider:
+			for sntoc := range s.snapsToConsiderCh {
+				select {
+				case <-stopCh:
+					break Consider
+				default:
+				}
+				var seedSnap *Snap
+				modes := runMode
+				essential := false
+				if sntoc.modelSnap != nil {
+					modes = sntoc.modelSnap.Modes
+					essential = sntoc.essential
+				}
+				if essential {
+					seedSnap = cachedEssential(sntoc.modelSnap.SnapType)
+				}
+				if seedSnap == nil {
+					var err error
+					seedSnap, err = s.doLoadMetaOne(&sntoc, jtm)
+					if err != nil {
+						if err == errSkipped {
+							continue
+						}
+						outcomesCh <- err
+						return
+					}
+					if essential {
+						cacheEssential(sntoc.modelSnap.SnapType, seedSnap)
+					}
+				}
+				i := sntoc.i
+				s.snaps[i] = seedSnap
+				s.modes[i] = modes
+			}
+			outcomesCh <- nil
+		}()
+	}
+	var firstErr error
+	done := 0
+	for done != njobs {
+		err := <-outcomesCh
+		done++
+		if err != nil && firstErr == nil {
+			firstErr = err
+			close(stopCh)
+		}
+	}
+	s.snapsToConsiderCh = nil
+	if firstErr != nil {
+		return firstErr
+	}
+	osnaps := s.snaps
+	omodes := s.modes
+	s.snaps = s.snaps[:0]
+	s.modes = s.modes[:0]
+	for i, sn := range osnaps {
```

Comment:
Is this only to remove nils if there are any? Though if doLoadMetaOne() did not fail, there would be no nils unless the snap was skipped right?

Reply:
yes, but we can have skipped snaps each time we filter by mode